File: WorldFeed/Client/src/app/core/Public/Parameters/TweetsClient/DestroyRetweetParameters.ts
import {CustomRequestParameters, ICustomRequestParameters} from "../CustomRequestParameters";
import {ITweetModeParameter} from "../ITweetModeParameter";
import {ITweetIdentifier} from "../../Models/Interfaces/ITweetIdentifier";
import {TweetIdentifier} from "../../Models/TweetIdentifier";
import {TweetMode} from '../../Settings/TweetinviSettings';

// For more information visit : https://dev.twitter.com/en/docs/tweets/post-and-engage/api-reference/post-statuses-unretweet-id
export interface IDestroyRetweetParameters extends ICustomRequestParameters, ITweetModeParameter {
  // The identifier of the retweet
  Tweet: ITweetIdentifier;

  // Tweets author object will not be populated when set to true
  TrimUser?: boolean;
}

export class DestroyRetweetParameters extends CustomRequestParameters implements IDestroyRetweetParameters {
  constructor(tweetIdOrTweet: | number | ITweetIdentifier) {
    super();

    let tweetCurrent: ITweetIdentifier;
    if (tweetIdOrTweet instanceof TweetIdentifier) {
      tweetCurrent = tweetIdOrTweet;
    } else if (typeof tweetIdOrTweet === 'number') {
      tweetCurrent = new TweetIdentifier(tweetIdOrTweet);
    }

    this.Tweet = tweetCurrent;
  }

  public Tweet: ITweetIdentifier;
  public TrimUser?: boolean;
  public TweetMode?: TweetMode;
}

// public DestroyRetweetParameters(long tweetId)
// {
//   Tweet = new TweetIdentifier(tweetId);
// }
//
// public DestroyRetweetParameters(ITweetIdentifier tweet)
// {
//   Tweet = tweet;
// }

File: Sharebook/Server/src/Services/Post/Sharebook.Post.API/Parameters/TweetsClient/DestroyRetweetParameters.cs
namespace Sharebook.Post.Application.Parameters.TweetsClient
{
    using Sharebook.Common.Public.Models;
    using Sharebook.Common.Public.Models.Interfaces;
    using Sharebook.Common.Public.Parameters;
    using Sharebook.Common.Settings;

    /// <summary>
    /// For more information visit : https://dev.twitter.com/en/docs/tweets/post-and-engage/api-reference/post-statuses-unretweet-id
    /// </summary>
    public interface IDestroyRetweetParameters : ICustomRequestParameters
    {
        /// <summary>
        /// The identifier of the retweet
        /// </summary>
        ITweetIdentifier Tweet { get; set; }

        /// <summary>
        /// Tweets author object will not be populated when set to true
        /// </summary>
        bool? TrimUser { get; set; }

        /// <summary>
        /// Decide whether to use Extended or Compat mode
        /// </summary>
        TweetMode? TweetMode { get; set; }
    }

    /// <inheritdoc/>
    public class DestroyRetweetParameters : CustomRequestParameters, IDestroyRetweetParameters
    {
        public DestroyRetweetParameters(long tweetId)
        {
            Tweet = new TweetIdentifier(tweetId);
        }

        public DestroyRetweetParameters(ITweetIdentifier tweet)
        {
            Tweet = tweet;
        }

        /// <inheritdoc/>
        public ITweetIdentifier Tweet { get; set; }
        /// <inheritdoc/>
        public bool? TrimUser { get; set; }
        /// <inheritdoc/>
        public TweetMode? TweetMode { get; set; }
    }
}
